packages/dds/tree/src/test/shared-tree/sharedTree.bench.ts

```diff
@@ -0,0 +1,739 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { strict as assert } from "assert";
+import { IRandom, makeRandom } from "@fluid-internal/stochastic-test-utils";
+import { benchmark, BenchmarkType } from "@fluid-tools/benchmark";
+import {
+	createField,
+	FieldKinds,
+	getField,
+	isUnwrappedNode,
+	namedTreeSchema,
+	singleTextCursor,
+} from "../../feature-libraries";
+import { brand, unreachableCase } from "../../util";
+import { ITestTreeProvider, TestTreeProvider } from "../utils";
+import { ISharedTree } from "../../shared-tree";
+import {
+	schemaMap,
+	// eslint-disable-next-line import/no-internal-modules
+} from "../feature-libraries/editable-tree/mockData";
+import {
+	FieldKindIdentifier,
+	fieldSchema,
+	GlobalFieldKey,
+	JsonableTree,
+	LocalFieldKey,
+	moveToDetachedField,
+	rootFieldKey,
+	rootFieldKeySymbol,
+	SchemaData,
+	TransactionResult,
+	TreeSchemaIdentifier,
+	UpPath,
+	ValueSchema,
+} from "../../core";
+
+enum TreeShape {
+	Wide = 0,
+	Deep = 1,
+}
+
+enum TestPrimitives {
+	Number = 0,
+	Float = 1,
+	String = 2,
+	Boolean = 3,
+	Map = 4,
+}
+
+async function createSharedTrees(
+	schemaData: SchemaData,
+	data: JsonableTree[],
+	numberOfTrees = 1,
+): Promise<readonly [ITestTreeProvider, readonly ISharedTree[]]> {
+	const provider = await TestTreeProvider.create(numberOfTrees);
+	for (const tree of provider.trees) {
+		assert(tree.isAttached());
+	}
+	provider.trees[0].storedSchema.update(schemaData);
+	provider.trees[0].context.root.insertNodes(0, data.map(singleTextCursor));
+	await provider.ensureSynchronized();
+	return [provider, provider.trees];
+}
+
+const globalFieldKey: GlobalFieldKey = brand("foo");
+const localFieldKey: LocalFieldKey = brand("foo");
+const rootSchemaName: TreeSchemaIdentifier = brand("Test");
+
+function getTestSchema(fieldKind: { identifier: FieldKindIdentifier }): SchemaData {
+	const testRootNodeSchema = namedTreeSchema({
+		name: rootSchemaName,
+		localFields: {
+			[localFieldKey]: fieldSchema(fieldKind),
+		},
+		globalFields: [globalFieldKey],
+		value: ValueSchema.Serializable,
+	});
+	schemaMap.set(rootSchemaName, testRootNodeSchema);
+	schemaMap.set(dataSchema.name, dataSchema);
+	return {
+		treeSchema: schemaMap,
+		globalFieldSchema: new Map([
+			[rootFieldKey, fieldSchema(FieldKinds.optional, [rootSchemaName])],
+			[globalFieldKey, fieldSchema(fieldKind)],
+		]),
+	};
+}
+
+const dataSchema = namedTreeSchema({
+	name: brand("DataSchema"),
+	localFields: {
+		foo: fieldSchema(FieldKinds.optional),
+	},
+	value: ValueSchema.Number,
+});
+
+const wideTestSizes = [5, 100, 500];
+const deepTestSizes = [5, 10, 100];
```

Comment:
Why did you pick these specific sizes? 5 nodes is a tree so small it should only be useful for measuring constant factors and we don't care about scaling that small, so it seems like "1" would make more sense for a minimum.

If the middle size is just for seeing if the large size scales linearly, might make more sense to mark it as BenchmarkType.Perspective

Also all the direct JS versions should be BenchmarkType.Perspective because they don't actually measure the performance of our system (so them lowing down is not a regression in tree).

Making less if the benchmarks BenchmarkType.Measurement means that when trying to measure our system performance running the BenchmarkType.Measurement subset of the suite will be faster and more meaningful (since it would be diluted by averaging in other benchmarks).

Reply:
updated sizes and benchmark types